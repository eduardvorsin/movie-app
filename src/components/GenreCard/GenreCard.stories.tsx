import type { Meta, StoryObj } from '@storybook/react';
import GenreCard from './GenreCard';
import { Genres } from '@/types/shared';

const genres: Genres[] = ['adventure', 'romance', 'history', 'music', 'war', 'thriller', 'animation', 'crime', 'documentary', 'drama', 'family', 'fantasy', 'horror', 'mystery', 'tv movie', 'western', 'action', 'comedy', 'action & adventure', 'war & politics', 'talk', 'soap', 'news', 'reality', 'kids', 'sci-fi & fantasy', 'science fiction'];

const meta: Meta<typeof GenreCard> = {
	title: 'components/GenreCard',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for GenreCard'
		},
		href: {
			description: 'The url to which the click-through will take place',
		},
		genre: {
			description: 'Type of genre',
		},
		title: {
			description: 'Title content',
		},
		titleElement: {
			description: 'Title level from 1 to 6',
		},
		onClick: {
			control: false,
			description: 'The callback that is called when you click on the card',
		},
		testId: {
			description: 'Id for testing the component',
		}
	},
	component: GenreCard,
	parameters: {
		docs: {
			description: {
				component: 'GenreCard is a type of cards that helps the user to select the desired movie genre for further searchreCards are employed to exhibit the condition of an object or the outcome of a completed action.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof GenreCard>;

export const Default: Story = {
	args: {
		title: 'Default',
		titleElement: 'h3',
		genre: 'comedy',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Default type of card.',
			},
		},
	},
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
	},
	args: {
		title: 'Click event',
		titleElement: 'h3',
		genre: 'adventure',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'Callback function can be supplied to onClick event',
			},
		},
	},
};

export const All: Story = {
	render: () => {
		return (
			<div
				className='grid grid-cols-[repeat(auto-fill,_150px)] auto-rows-min gap-4'
			>
				{genres.map((genre) => (
					<GenreCard
						key={genre}
						genre={genre}
						title={`${genre} title`}
						titleElement='h3'
						href='/'
					/>
				))}
			</div>
		)
	},
	parameters: {
		docs: {
			description: {
				story: 'Here are all possible variants of GenreCards by types of icons and genres',
			},
		},
	},
};