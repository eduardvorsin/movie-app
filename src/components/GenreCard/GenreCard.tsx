'use client';
import Link from 'next/link';
import { Adventure, Romance, Fantastic, History, Music, War, Thriller, Animation, Crime, Documentary, Drama, Family, Fantasy, Horror, Mystery, TVMovie, Western, Action, Comedy, ActionAndAdventure, Politics, Talk, Series, News, Reality, Kids, SciFi, Science } from './GenreIcons';
import { MouseEventHandler } from 'react';
import Title from '@/components/Title/Title';
import { GeneralProps, Genres, HeadingElement } from '@/types/shared';

const iconsByGenre = {
	adventure: {
		name: 'adventure',
		src: Adventure.src,
	},
	romance: {
		name: 'romance',
		src: Romance.src,
	},
	fantastic: {
		name: 'fantastic',
		src: Fantastic.src,
	},
	history: {
		name: 'history',
		src: History.src,
	},
	music: {
		name: 'music',
		src: Music.src,
	},
	war: {
		name: 'war',
		src: War.src,
	},
	thriller: {
		name: 'thriller',
		src: Thriller.src,
	},
	animation: {
		name: 'animation',
		src: Animation.src,
	},
	crime: {
		name: 'crime',
		src: Crime.src,
	},
	documentary: {
		name: 'documentary',
		src: Documentary.src,
	},
	drama: {
		name: 'drama',
		src: Drama.src,
	},
	family: {
		name: 'family',
		src: Family.src,
	},
	fantasy: {
		name: 'fantasy',
		src: Fantasy.src,
	},
	horror: {
		name: 'horror',
		src: Horror.src,
	},
	mystery: {
		name: 'mystery',
		src: Mystery.src,
	},
	'tv movie': {
		name: 'tv',
		src: TVMovie.src,
	},
	western: {
		name: 'western',
		src: Western.src,
	},
	action: {
		name: 'action',
		src: Action.src,
	},
	comedy: {
		name: 'comedy',
		src: Comedy.src,
	},
	'action & adventure': {
		name: 'action_adventure',
		src: ActionAndAdventure.src,
	},
	'war & politics': {
		name: 'politics',
		src: Politics.src,
	},
	talk: {
		name: 'talk',
		src: Talk.src,
	},
	soap: {
		name: 'series',
		src: Series.src,
	},
	news: {
		name: 'news',
		src: News.src,
	},
	reality: {
		name: 'reality',
		src: Reality.src,
	},
	kids: {
		name: 'kids',
		src: Kids.src,
	},
	'sci-fi & fantasy': {
		name: 'sci-fi',
		src: SciFi.src,
	},
	'science fiction': {
		name: 'science',
		src: Science.src,
	},
} as const;

export type Props = {
	href: string,
	genre: Genres,
	title: string,
	titleElement: HeadingElement,
	onClick?: MouseEventHandler<HTMLAnchorElement>,
} & GeneralProps;

export default function GenreCard({
	href,
	className,
	genre,
	title,
	titleElement,
	onClick,
	testId,
	...props
}: Props) {
	const classes = [
		'px-2 py-6 rounded-2 h-[6.75rem] w-[9.75rem] bg-neutral-200 dark:bg-dark-neutral-300 [&]:text-neutral-1000 dark:[&]:text-dark-neutral-900 flex flex-col items-center relative after:absolute after:top-0 after:left-0 after:z-100 after:w-full after:h-full transition-colors duration-150 hover:bg-neutral-300 active:bg-neutral-400 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500 ',
		className,
	].join(' ');

	return (
		<Link
			className={classes}
			href={href}
			onClick={onClick}
			data-testid={testId}
			{...props}
		>
			<svg className='w-8 h-8 mb-2 fill-current' viewBox='0 0 32 32'>
				<use href={`${iconsByGenre[genre].src}#${iconsByGenre[genre].name}`}></use>
			</svg>
			<Title
				level={6}
				weight={700}
				as={titleElement}
				className='leading-2 text-center w-full truncate'
			>
				{title}
			</Title>
		</Link>
	);
};
