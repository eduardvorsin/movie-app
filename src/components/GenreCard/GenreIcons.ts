export { default as Adventure } from '../../assets/icons/adventure.svg?url';
export { default as Romance } from '../../assets/icons/hearts.svg?url';
export { default as Fantastic } from '../../assets/icons/ufo.svg?url';
export { default as History } from '../../assets/icons/history.svg?url';
export { default as Music } from '../../assets/icons/music.svg?url';
export { default as Military } from '../../assets/icons/military.svg?url';
export { default as Thriller } from '../../assets/icons/thriller.svg?url';
export { default as Animation } from '../../assets/icons/kid.svg?url';
export { default as Crime } from '../../assets/icons/crime.svg?url';
export { default as Documentary } from '../../assets/icons/documentary.svg?url';
export { default as Drama } from '../../assets/icons/drama.svg?url';
export { default as Family } from '../../assets/icons/family.svg?url';
export { default as Fantasy } from '../../assets/icons/fantasy.svg?url';
export { default as Horror } from '../../assets/icons/horror.svg?url';
export { default as Mystery } from '../../assets/icons/mystery.svg?url';
export { default as TVMovie } from '../../assets/icons/tv.svg?url';
export { default as Western } from '../../assets/icons/western.svg?url';
export { default as Action } from '../../assets/icons/action.svg?url';
export { default as Comedy } from '../../assets/icons/comedy.svg?url';
