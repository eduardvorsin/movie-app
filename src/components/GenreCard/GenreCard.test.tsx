import { render, screen } from '@testing-library/react';
import GenreCard from './GenreCard';
import userEvent from '@testing-library/user-event';

describe('GenreCard tests', () => {
	it('is rendered correctly', () => {
		render(
			<GenreCard
				href='/'
				genre='comedy'
				title='comedy'
				titleElement='h3'
			/>
		);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toBeInTheDocument();
	});

	it('when you click on the link, the mock function should call', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<GenreCard
				href='/'
				genre='horror'
				title='horror'
				titleElement='h3'
				onClick={mockFn}
			/>
		);

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		render(
			<GenreCard
				href='/'
				genre='history'
				title='history'
				titleElement='h3'
			/>
		);

		expect(screen.getByRole<HTMLAnchorElement>('link')).toMatchSnapshot();
	});
});