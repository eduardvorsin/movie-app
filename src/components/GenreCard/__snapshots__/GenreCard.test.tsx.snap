// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreCard tests is a basic snapshot 1`] = `
<a
  class="px-2 py-6 rounded-2 h-[6.75rem] w-[9.75rem] bg-neutral-200 dark:bg-dark-neutral-300 text-neutral-1000 dark:text-dark-neutral-900 flex flex-col items-center relative after:absolute after:top-0 after:left-0 after:z-100 after:w-full after:h-full transition-colors duration-150 hover:bg-neutral-300 active:bg-neutral-400 dark:hover:bg-dark-neutral-400 dark:active:bg-dark-neutral-500  "
  href="/"
>
  <svg
    class="w-8 h-8 mb-2 fill-current"
    viewBox="0 0 32 32"
  >
    <use
      href="undefined#history"
    />
  </svg>
  <h3
    class="-tracking-[0.01em] text-200 leading-1 font-bold leading-2 text-center w-full truncate"
  >
    history
  </h3>
</a>
`;
